Rework the receipts storage in executor aux db

liuchengxu opened the pull request:
This PR is initiated to resolve a bug #1353 on gemini-3d, caused by the collision that multiple primary blocks can produce one identical domain block. The key change here is to track the execution receipts in the executor local database using the `primary_hash` as the key instead of the current `domain_hash`, which is also related to #1254 in which we found using `primary_hash` as the key is generally a better option as well.

A few TODOs along the way are also resolved in this PR, one of them is to ensure the returned first unconfirmed bad receipt still points to a block on the canonical chain, ~~it's now done by traversing from the tip to the target, which can be improved once we have a better answer [here](https://substrate.stackexchange.com/questions/7970/whats-the-best-practice-to-check-whether-a-block-hash-is-in-the-canonical-chain). I may dive into the code later if no good answer from the expert.~~ (fixed)

A test is added to ensure the collected receipts are on the same branch with the current best primary block when producing a bundle, even when having multiple primary chain forks.

Executors on gemini-3d need to purge the data before running the next release with this patch.

Close #1353 
Close #1254

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

liuchengxu committed:
Track the receipt in executor aux db using `primary_hash` as key

liuchengxu committed:
Integrate `primary_chain_client` into DomainBundleProposer

Prepare for retrieving the primary block hash using the number.

liuchengxu committed:
Track the mapping of (domain_hash, primary_hash) in aux db

Upon the incoming primary block notification, we need the hash of
primary block that derives the current domain block in order to
calculate the primary chain tree route.

It was retrieved from the receipt related to the best domain block
before, now this mapping has to be stored separately in the aux db
due to the change of receipts tracking, it will be pruned periodically
later.

liuchengxu committed:
Update the read of receipt and disable the related inactive code

Now the receipt is loaded from aux db using the key `primary_block_hash`.

The receipt reading is involved on a few inactive code path, I disabled
them for simplicity in order to make it compile.

liuchengxu committed:
Ensure the first unconfirmed bad receipt still points to a block on the canonical chain

The way of checking whether a block is on the canonical chain may be
improved later, it's now using a way conservative but correct.

liuchengxu committed:
Resolve fixed TODOs

- Checking whether the bad receipt points to a canonical block is done
  in the last commit.
- Regarding the bad fork receipts, I realize there is no need to handle
  them separately because all the bad receipts will be pruned later once
  they are older than `oldest_receipt_number`, see
  `crate::aux_schema::prune_expired_bad_receipts()`.

liuchengxu committed:
Fix `trace_mismatch_index` finding

Close https://github.com/subspace/subspace/issues/1254

liuchengxu committed:
Enrich the debug info of submitted bundle

liuchengxu committed:
Add log prefix to mock primary node and use longest chain rule simply

There are also several minor refactorings for writing test in the next
commit.

liuchengxu committed:
Add test `collected_receipts_should_be_on_the_same_branch_with_current_best_block`

This test ensures the receipts in the bundle are always on the same
branch with current best primary block.

liuchengxu committed:
Fix test `pallet_domains_unsigned_extrinsics_should_work`

liuchengxu committed:
Check whether the receipt is available before producing a slot in test

Fix https://github.com/subspace/subspace/actions/runs/4648954717/jobs/8226975359

vedhavyas commented on domains/client/domain-executor/src/aux_schema.rs:
```diff
@@ -128,6 +140,39 @@ where
     load_decode(backend, execution_receipt_key(block_hash).as_slice())
 }
 
+pub(super) fn track_domain_hash_to_primary_hash<Backend, Hash, PHash>(
```
why track them when you can already get them from the runtime?

liuchengxu replied:
I don't know it's possible to get the corresponding primary block hash given the domain block hash from the runtime. Are you sure about that?

vedhavyas replied:
I was mostly referring to this function 
https://github.com/subspace/subspace/blob/main/crates/pallet-receipts/src/lib.rs#L237
if we modify this and accept an execution receipt hash, we can fetch that, but I see that would need some work.
Ignore my suggestion if this is unnecessary 👍🏼

liuchengxu replied:
Relying on the receipt from the runtime requires the corresponding receipt is able to be submitted successfully in time (it may fail due to some bug as we have already seen :D) otherwise the domain block processing will be blocked, we probably don't want that.

vedhavyas reviewed (commented):
Overall change make sense to me

liuchengxu committed:
Improve canonical primary block hash retrieval

As as matter of fact, `client.hash(number)` returns the canonical block
hash for given block number.

NingLin-P commented on domains/client/domain-executor/src/tests.rs:
```diff
@@ -95,6 +95,40 @@ async fn collected_receipts_should_be_on_the_same_branch_with_current_best_block
+
+    assert!(
+        load_receipt(best_primary_hash).is_some(),
+        "Executor failed to process and generate the receipt of last primary block in 50ms"
```
This may make the test unstable (depending on whether the CI is slow or not), how about calling `primary_node .notify_new_slot_and_wait_for_bundle` here. When the slot handling is finished, the previous block import task must be finished.

liuchengxu replied:
I don't see how `primary_node.notify_new_slot_and_wait_for_bundle` would help, the bundle production and block processing run concurrently. As I have noted in the comments above, the domain block import notification does not mean the completion of a domain block pipeline. The potential issue here is when producing a bundle on a new slot, the receipt is likely not yet available and then it fails at this slot https://github.com/subspace/subspace/actions/runs/4648954717/jobs/8226975359#step:7:963, what we want to is to always succeed in producing this bundle.

NingLin-P replied:
Sorry, I mixed them up, but originally this situation should not happen.

Because, in the block import pipeline,  the `blocks_imported` notification actually comes before `blocks_importing` because `inner.import_block` (namely the `client.import_block` which sends the `every_import_notification` notification) is called before `acknowledgement_sender` is sent. 

And as `ferdie.produce_blocks(1)` will wait for the acknowledgment before completing, the `blocks_imported` notification must be handled since it is sent before `acknowledgement_sender` and we have set `primary_block_import_throttling_buffer_size` to 0 in the test.

But there are two places we did not consider:
- `tokio::select` randomly picks a branch to check first (see https://docs.rs/tokio/latest/tokio/macro.select.html#fairness), which means when both `every_import_notification` and `acknowledgement_sender` are sent, it will randomly pick one to handle instead of based on the order they comes.
- `futures::channel::mpsc::channel` actually have `buffer + num-senders` number of slot (see https://docs.rs/futures/latest/futures/channel/mpsc/fn.channel.html), which means even we have set `primary_block_import_throttling_buffer_size` to 0, it still has 1 slot thus acknowledgement may return before `blocks_imported` notification is handled

To fix that we may need to:
- Add `biased;` to the beginning of the `tokio::select` usage as the document suggested to ensure the order of branches to process, this should not affect the production code as we didn't have any assumption on this.
- Send one more `acknowledgement_sender` and wait for one more acknowledgement in the mock primary node, when this acknowledgement returns we can ensure `blocks_imported` notification must be handled.

WDYT @liuchengxu

liuchengxu replied:
This sounds interesting, the reason now we have the term `ImportedBlock` and `BlockImporting` is in that we did run into #1143 which looked like the block is not fully imported at which point the notification from sc-consensus-subspace is fired, it's contrary to the analysis here (it makes some sense to me actually) that the notification of `ImportedBlock` is fired before `BlockImporting` in theory, if it does work as this way, we shouldn't encounter #1143. I suspect something from substrate or elsewhere I can't tell does not work as expected.

- Adding `biased` is okay to me, we won't set the buffer size to zero except for the tests anyway.
- I'm not sure about sending another ack. Having the buffer size as 0 and one sender means the channel will be blocked unless each received message is processed in my understanding. If the `ImportedBlock` notification always comes before `BlockImporting` with `biased`, then the corresponding `BlockImporing` notification won't be received until the notified primary block is fully processed by the executor, isn't this enough?

How about leaving this workaround as is so that this PR can move forward as it should be practically helpful to some extent IMO? And file a new dedicated as I believe there are multiple places that seemingly are affected, e.g., https://github.com/subspace/subspace/actions/runs/4657062650/jobs/8241286079?pr=1365#step:7:986,

NingLin-P replied:
> if it does work as this way, we shouldn't encounter https://github.com/subspace/subspace/issues/1143. I suspect something from substrate or elsewhere I can't tell does not work as expected

Yeah, I feel the same... but I have added `biased` and some logs and then run a test multiple times, which always shows `ImportedBlock` is fired before `BlockImporting` thus I think we are good to proceed this step with the test.

> then the corresponding BlockImporing notification won't be received until the notified primary block is fully processed by the executor

It may not, `futures::channel::mpsc::channel(0)` can send 1 message without blocking, thus after `ImportedBlock ` is **received**, `BlockImporing` can be sent without blocking (so the `acknowledgement_sender` will also respond), and at the time `ImportedBlock` may not be fully processed.

> How about leaving this workaround as is so that this PR can move forward as it should be practically helpful to some extent IMO? And file a new dedicated as I believe there are multiple places that seemingly are affected

Sure! I will open a dedicated PR for the abovementioned changes.

NingLin-P reviewed (commented):
Mostly make sense to me.

NingLin-P reviewed (commented):
Mostly make sense to me.

liuchengxu committed:
Merge branch 'main' into change-receipt-key-to-primary-block-hash
